### PR TITLE
test: fixing waitforaccesslogs

### PR DESCRIPTION
**Pull request by alyssawilk**
It silently returns multiple access logs, making some tests not test what they indicated they tested.

Risk Level: low (test only)
Testing: yes
Docs Changes: n/a
Release Notes: n/a
related to #21410

### CODE REVIEW

File: test/integration/base_integration_test.h [hunk 1]
```diff
@@ -146,8 +146,10 @@ class BaseIntegrationTest : protected Logger::Loggable<Logger::Id::testing> {
 
   // Enable the listener access log
   void useListenerAccessLog(absl::string_view format = "");
-  // Waits for the nth access log entry, defaulting to log entry 0.
-  std::string waitForAccessLog(const std::string& filename, uint32_t entry = 0);
+  // Returns all log entries after the nth access log entry, defaulting to log entry 0.
+  // By default will assert fail if more than one entry is returned.
```
Comment (KBaichoo) [inline test/integration/base_integration_test.h:150]: we assert fail, but we use EXPECT macros?

**Review by KBaichoo — COMMENTED** [review]
looks good otherwise, thanks!

File: test/integration/base_integration_test.cc [hunk 2]
```diff
@@ -482,13 +482,18 @@ size_t entryIndex(const std::string& file, uint32_t entry) {
...
+std::string BaseIntegrationTest::waitForAccessLog(const std::string& filename, uint32_t entry,
+                                                  bool allow_excess_entries) {
   // Wait a max of 1s for logs to flush to disk.
   std::string contents;
   for (int i = 0; i < 1000; ++i) {
     contents = TestEnvironment::readFileToStringForTest(filename);
     size_t index = entryIndex(contents, entry);
     if (contents.length() > index) {
+      if (!allow_excess_entries) {
+        EXPECT_EQ(contents.length(), entryIndex(contents, entry + 1))
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply test/integration/base_integration_test.cc:494]: given the common case is less than 100 characters and 2 log entries, I don't think I mind walking it twice.  I mean it's also 2N which is still O(n) so I'm inclined towards simplicity unless you punt it back at me in which case I'll do it :-P

File: test/extensions/filters/http/cache/cache_filter_integration_test.cc [hunk 3]
```diff
@@ -276,7 +276,7 @@ TEST_P(CacheIntegrationTest, ExpiredFetchedNewResponse) {
 
     // Advance time to force a log flush.
     simTime().advanceTimeWait(Seconds(1));
-    EXPECT_THAT(waitForAccessLog(access_log_name_), testing::HasSubstr("- via_upstream"));
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply test/extensions/filters/http/cache/cache_filter_integration_test.cc:279]: before we were waiting for entry 0 when we wanted 1.  so it wasn't actually checking for the second response.

**Review by KBaichoo — APPROVED** [review]
lgtm, thanks!

**Pull request merged**